new formula: otf2 2.1.1

XL64 opened the pull request:
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

XL64 commented:
The OTF2 team suggest me to insert a make check between make and make install.
That would be more relevant than doing the simple tests that are in the test section of the recipe.
I don't know if its the homebrew policy to run the check step or not.

equal-l2 commented:
The `test do` block can be run by users to test the bottle is successfully poured.
It is fine to run `make check` in `install` block, but you are definitely required to implement the `test do` block, too.

XL64 commented:
Added make check then thanks.

XL64 commented:
Before merging, I was advice to add a python dependency by the development team (for python bindings), I'll add it

SMillerDev commented on Formula/otf2.rb:
```diff
@@ -0,0 +1,29 @@
+                          "--disable-dependency-tracking",
+                          "--disable-silent-rules",
+                          "--prefix=#{prefix}",
+                          "--enable-backend-test-runs"
```
What's the impact of this on the resulting binary?

XL64 replied:
As far as I understand, this allows one to run some testing on the library built using make check.

SMillerDev replied:
Not sure if it'd be good to have such debug options in the build we ship out to users.

XL64 replied:
Ok I'll remove and push

XL64 replied:
done

zbeekman commented on Formula/otf2.rb:
```diff
@@ -0,0 +1,27 @@
+      # build serial tests
+      system "make", "serial"
+      system "./otf2_writer_example"
+      system "./otf2_reader_example"
```
It would be good to verify output from running commands. I guess if the reader reads the example that the writer wrote this might be ok. But it would be good to test for the creation of expected files and/or expected strings in stdout indicating success.

zbeekman replied:
@XL64 Any chance you could update the tests with additional checks?

zbeekman replied:
Let's also add the MPI checks too, and Fortran.

zbeekman commented:
@BrewTestBot test this please

XL64 commented:
The writter test data are used by the reader as far as I understand. But I
don't know what to test for...

Le mer. 10 avr. 2019 à 15:22, zbeekman <notifications@github.com> a écrit :

> *@zbeekman* commented on this pull request.
> ------------------------------
>
> In Formula/otf2.rb
> <https://github.com/Homebrew/homebrew-core/pull/37719#discussion_r273955377>
> :
>
> > +  def install
> +    system "./configure", "--disable-debug",
> +                          "--disable-dependency-tracking",
> +                          "--disable-silent-rules",
> +                          "--prefix=#{prefix}"
> +    system "make"
> +    system "make", "install"
> +  end
> +
> +  test do
> +    cp_r "#{share}/doc/otf2/examples", testpath
> +    chdir "#{testpath}/examples" do
> +      # build serial tests
> +      system "make", "serial"
> +      system "./otf2_writer_example"
> +      system "./otf2_reader_example"
>
> @XL64 <https://github.com/XL64> Any chance you could update the tests
> with additional checks?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Homebrew/homebrew-core/pull/37719#discussion_r273955377>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABfOMjwj1f0swFzsLMWwJHDHSept-OSGks5vfeWFgaJpZM4blVol>
> .
>

zbeekman commented:
These tests are insufficient. Running with `brew test --verbose` shows multiples errors:

```
==> make serial
gcc -std=c99 `otf2-config --cflags` \
            -c otf2_reader_example.c \
            -o otf2_reader_example.o
gcc -std=c99 `otf2-config --cflags` \
            -c otf2_writer_example.c \
            -o otf2_writer_example.o
gcc otf2_writer_example.o \
            `otf2-config --ldflags` \
            `otf2-config --libs` \
            -o otf2_writer_example
gcc otf2_reader_example.o \
            `otf2-config --ldflags` \
            `otf2-config --libs` \
            -o otf2_reader_example
==> ./otf2_writer_example
==> ./otf2_reader_example
[OTF2] src/otf2_file_posix.c:163: error: File or directory does not exist: POSIX: 'ArchivePath/ArchiveName/0.def'
[OTF2] src/OTF2_Buffer.c:2047: error: File or directory does not exist: Could not open file.
[OTF2] src/OTF2_Buffer.c:2078: error: File or directory does not exist: Failed to get file handle!
[OTF2] src/OTF2_Buffer.c:586: error: File or directory does not exist: Could not read data from file to buffer!
[OTF2] src/OTF2_Archive.c:1441: error: File or directory does not exist: Could not get local definition reader
Entering region 0 at location 0 at time 0.
Leaving region 0 at location 0 at time 1.
```

Not only are these tests silently failing, but they need to expanded. You can detect correct and incorrect output, or the existence of files.

XL64 commented:
@zbeekman 
I already noticed those output and reported to the development team.
Here is the answer I got:

>  the return code from otf2_reader_example should nevertheless be 'success'. If so, the example works. The error messages only correspond to the optional 'local definitions'.

I agree these messages are misleeading but the run is correct...

zbeekman commented on Formula/otf2.rb:
```diff
@@ -0,0 +1,27 @@
+  url "https://www.vi-hps.org/cms/upload/packages/otf2/otf2-2.1.1.tar.gz"
+  sha256 "01591b42e76f396869ffc84672f4eaa90ee8ec2a8939755d9c0b5b8ecdcf47d3"
+
+  depends_on "python"
```
```suggestion
  depends_on "gcc"
  depends_on "open-mpi"
  depends_on "python"
```

Should build for Fortran support and MPI support too. I'm not so sure about python here;

I think without the explicit python dependency it should pick up the system python and build bindings for that.

XL64 commented:
That makes sense. Should I amend my pull request ?

Le ven. 12 avr. 2019 à 20:37, zbeekman <notifications@github.com> a écrit :

> *@zbeekman* commented on this pull request.
> ------------------------------
>
> In Formula/otf2.rb
> <https://github.com/Homebrew/homebrew-core/pull/37719#discussion_r275020533>
> :
>
> > @@ -0,0 +1,27 @@
>
> +class Otf2 < Formula
>
> +  desc "Open Trace Format 2 file handling library"
>
> +  homepage "https://www.vi-hps.org/projects/score-p/"
>
> +  url "https://www.vi-hps.org/cms/upload/packages/otf2/otf2-2.1.1.tar.gz"
>
> +  sha256 "01591b42e76f396869ffc84672f4eaa90ee8ec2a8939755d9c0b5b8ecdcf47d3"
>
> +
>
> +  depends_on "python"
>
>
> ⬇️ Suggested change
>
> -  depends_on "python"
>
> +  depends_on "gcc"
>
> +  depends_on "open-mpi"
>
> +  depends_on "python"
>
>
> Should build for Fortran support and MPI support too. I'm not so sure
> about python here;
>
> I think without the explicit python dependency it should pick up the
> system python and build bindings for that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Homebrew/homebrew-core/pull/37719#pullrequestreview-226230883>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABfOMpQMpBx9_5_qqMmhaMjillGsGHWtks5vgNJjgaJpZM4blVol>
> .
>

XL64 commented:
My goal was tout build cube and thus i didn't need mpi support but Indeed
we could do that.

Le ven. 12 avr. 2019 à 20:38, zbeekman <notifications@github.com> a écrit :

> *@zbeekman* commented on this pull request.
> ------------------------------
>
> In Formula/otf2.rb
> <https://github.com/Homebrew/homebrew-core/pull/37719#discussion_r275020756>
> :
>
> > +  def install
> +    system "./configure", "--disable-debug",
> +                          "--disable-dependency-tracking",
> +                          "--disable-silent-rules",
> +                          "--prefix=#{prefix}"
> +    system "make"
> +    system "make", "install"
> +  end
> +
> +  test do
> +    cp_r "#{share}/doc/otf2/examples", testpath
> +    chdir "#{testpath}/examples" do
> +      # build serial tests
> +      system "make", "serial"
> +      system "./otf2_writer_example"
> +      system "./otf2_reader_example"
>
> Let's also add the MPI checks too, and Fortran.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Homebrew/homebrew-core/pull/37719#discussion_r275020756>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABfOMl7OW_IuPJe0TL1MUhHG1_POTVIHks5vgNKMgaJpZM4blVol>
> .
>

zbeekman commented:
> Before merging, I was advice to add a python dependency by the development team (for python bindings), I'll add it

FYI, this will add a Python3 dependency for this formula, Python2 is available and the formula should be building python 2 bindings (assuming the system python is up to date with the minimum version they require). Python3 is a very large dependency. I'm happy to keep it, but I just want to make sure that this really makes sense.

I'm going to add the Fortran support and MPI support and look into expanding the tests myself unless you beat me to it.

Also, it would be nice to look into which other performance analysis packages are supported on macOS that are not yet in Homebrew, Opari, PAPI, likwid, etc.

zbeekman commented:
Sure if you could amend and force push that would be great! I also want to
make sure that there are good/enough tests of the various capabilities. But
feel free to amend and force push.

On Fri, Apr 12, 2019 at 2:49 PM Xavier Lacoste <notifications@github.com>
wrote:

> My goal was tout build cube and thus i didn't need mpi support but Indeed
> we could do that.
>
> Le ven. 12 avr. 2019 à 20:38, zbeekman <notifications@github.com> a écrit
> :
>
> > *@zbeekman* commented on this pull request.
> > ------------------------------
> >
> > In Formula/otf2.rb
> > <
> https://github.com/Homebrew/homebrew-core/pull/37719#discussion_r275020756
> >
> > :
> >
> > > + def install
> > + system "./configure", "--disable-debug",
> > + "--disable-dependency-tracking",
> > + "--disable-silent-rules",
> > + "--prefix=#{prefix}"
> > + system "make"
> > + system "make", "install"
> > + end
> > +
> > + test do
> > + cp_r "#{share}/doc/otf2/examples", testpath
> > + chdir "#{testpath}/examples" do
> > + # build serial tests
> > + system "make", "serial"
> > + system "./otf2_writer_example"
> > + system "./otf2_reader_example"
> >
> > Let's also add the MPI checks too, and Fortran.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/Homebrew/homebrew-core/pull/37719#discussion_r275020756
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ABfOMl7OW_IuPJe0TL1MUhHG1_POTVIHks5vgNKMgaJpZM4blVol
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Homebrew/homebrew-core/pull/37719#issuecomment-482683868>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAREPM6HpI6FpI2BXkDXQSRrScNLCY90ks5vgNVSgaJpZM4blVol>
> .
>

zbeekman commented:
I've been digging around a bit more. Even when adding the python dependencies, the python bindings are not being built because of a missing `future` module, which is required by the configure script. I have also added some more thorough tests on my local copy of your branch. I'm investigating the right way to deal with the python module.

XL64 commented:
On my side, I undestood that the configure script will find system python rather that python3 installed with homebrew.
Moreover if I add python@2 cellar to the PATH it finds it but the builtins package is missing. I don't know if/how I can add a pip dependency to a homebrew recipe.
And when I do "/usr/local/Cellar/python\@2/2.7.16/bin/pip search builtins" I don't know which one to add...

XL64 commented:
As far as I understand, the future python module is required for the python binding. Thus do I need to add
```
  resource "future" do
    url "https://files.pythonhosted.org/packages/00/2b/8d082ddfed935f3608cc61140df6dcbf0edea1bc3ab52fb6c29ae3e81e85/future-0.16.0.tar.gz"
    sha256 "e39ced1ab767b5936646cedba8bcce582398233d6a627067d4c6a454c90cfedb"
  end
```
and
```
  def install
    virtualenv_install_with_resources
```
?

zbeekman commented:
@XL64 exactly, that's the route to go down. I'm a bit fuzzy on how to deal with the autotools build & install such that the "future" module is included and made available at build & runtime though.

XL64 commented:
I have no idea either... Do we really need the fortran bindings ?

XL64 commented:
I could run the following tests:

```

      system "make", "serial", "mpi"
      system "./otf2_writer_example"
      system "./otf2_reader_example"
      system "rm", "-rf", "ArchivePath"
      system "mpirun", "-np", "2", "./otf2_mpi_writer_example"
      system "mpirun", "-np", "2", "./otf2_mpi_reader_example"
```

Still the same expected "error" reported in brew test -v otf2... 
Is it better ?

zbeekman commented on Formula/otf2.rb:
```diff
@@ -0,0 +1,66 @@
+        assert_predicate workdir/"ArchivePath/ArchiveName/#{n}.evt", :exist?
+      end
+      system Formula["open-mpi"].opt_bin/"mpirun", "-n", "2", "./otf2_mpi_reader_example"
+      system "./otf2_reader_example"
```
It would be nice to ingest the stdout of the `reader_example` programs to verify correct functionality (and to throw out the warnings they print), but I think this is sufficient testing.

zbeekman commented:
Thanks for your contribution to Homebrew! 🎉 🥇 

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

XL64 commented:
Thanks @zbeekman. In fact I have a core recipe for cubegui because i wanted
to test experimental plug-in (blade). It worked if the trace is not too
large.
I didn't use scorep on Mac because i only profile codes on clusters...
Will see if i have Time to test those.


Le mer. 24 avr. 2019 à 21:22, zbeekman <notifications@github.com> a écrit :

> *@zbeekman* approved this pull request.
>
> @XL64 <https://github.com/XL64>: I added better tests and fixed it so
> that it will install Python bindings. I hope you don't mind that I force
> pushed to your fork/branch.
>
> It would be nice to also add CubeW, OPARI, and scorep. Would love to see
> those in a future PR if you have time.
>
> FYI I opened a PR to add CubeGUI to homebrew/cask.
> ------------------------------
>
> In Formula/otf2.rb
> <https://github.com/Homebrew/homebrew-core/pull/37719#discussion_r278279752>
> :
>
> > +        otf2_mpi_writer_example
> +        otf2_pthread_writer_example
> +        otf2_reader_example
> +        otf2_writer_example
> +      ].each { |p| assert_predicate workdir/p, :exist? }
> +      system "./otf2_writer_example"
> +      assert_predicate workdir/"ArchivePath/ArchiveName.otf2", :exist?
> +      system "./otf2_reader_example"
> +      rm_rf "./ArchivePath"
> +      system Formula["open-mpi"].opt_bin/"mpirun", "-n", "2", "./otf2_mpi_writer_example"
> +      assert_predicate workdir/"ArchivePath/ArchiveName.otf2", :exist?
> +      (0...2).each do |n|
> +        assert_predicate workdir/"ArchivePath/ArchiveName/#{n}.evt", :exist?
> +      end
> +      system Formula["open-mpi"].opt_bin/"mpirun", "-n", "2", "./otf2_mpi_reader_example"
> +      system "./otf2_reader_example"
>
> It would be nice to ingest the stdout of the reader_example programs to
> verify correct functionality (and to throw out the warnings they print),
> but I think this is sufficient testing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Homebrew/homebrew-core/pull/37719#pullrequestreview-230307433>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAL44MSCMSCUOGM2VPRQVOLPSCXP5ANCNFSM4G4VLISQ>
> .
>
